src/libAtomVM/nifs.c

```diff
@@ -1284,12 +1284,41 @@ static term nif_erlang_spawn(Context *ctx, int argc, term argv[])
 static term nif_erlang_send_2(Context *ctx, int argc, term argv[])
 {
     UNUSED(argc);
+    term target = argv[0];
+    GlobalContext *glb = ctx->global;
+
+    if (term_is_pid(target)) {
+        int32_t local_process_id = term_to_local_process_id(target);
+
+        globalcontext_send_message(glb, local_process_id, argv[1]);
+
+    } else if (term_is_atom(target)) {
+        // We need to hold a lock on the processes_table until the message is sent to avoid a race condition,
+        // otherwise the receiving process could be killed at any point between checking it is registered,
+        // obtaining its process_id, and sending the message, any of which would lead to crashes or incorrect
+        // behavior that does not match OTP.
+        struct ListHead *dummy = synclist_rdlock(&glb->processes_table);
+        UNUSED(dummy);
+        int atom_index = term_to_atom_index(target);
+
+        int target_id = globalcontext_get_registered_process(glb, atom_index);
+        if (UNLIKELY(target_id == 0)) {
+            synclist_unlock(&glb->processes_table);
+            RAISE_ERROR(BADARG_ATOM);
+        }
 
-    term pid_term = argv[0];
-    VALIDATE_VALUE(pid_term, term_is_pid);
+        Context *p = globalcontext_get_process_nolock(glb, target_id);
+        if (IS_NULL_PTR(p)) {
+            synclist_unlock(&glb->processes_table);
+            RAISE_ERROR(BADARG_ATOM);
+        }
 
-    int local_process_id = term_to_local_process_id(pid_term);
-    globalcontext_send_message(ctx->global, local_process_id, argv[1]);
+        int32_t local_process_id = p->process_id;
```

Comment:
I'm not yet understanding this: I suppose that `target_id` is always equals to `p->process_id` so I don't see any reason about about doing `int32_t local_process_id = p->process_id`.

As a further hint:
```c
static Context *globalcontext_get_process_nolock(GlobalContext *glb, int32_t process_id)
[...]
206         if (p->process_id == process_id) {
207             return p;
208         }
```

Reply:
You are absolutely correct, the extra assignment is unnecessary. I think this is an unnecessary hold-over from when I had `local_process_id` (or whatever variable name that replaced) declared at the top of the function, before branching...